crates/anvil-polkadot/src/api_server/txpool_helpers.rs

```diff
@@ -0,0 +1,212 @@
+//! Helper functions for txpool RPC methods
+//!
+//! This module contains utilities for extracting transaction information from
+//! Substrate extrinsics, including support for impersonated transactions with
+//! fake signatures.
+
+use alloy_primitives::{Address, B256, U256, keccak256};
+use alloy_rpc_types::txpool::TxpoolInspectSummary;
+use codec::{DecodeLimit, Encode};
+use polkadot_sdk::{
+    pallet_revive::evm::TransactionSigned,
+    sp_core::{self, H256},
+};
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use substrate_runtime::{RuntimeCall, UncheckedExtrinsic};
+
+use crate::substrate_node::host::is_impersonated;
+
+const MAX_EXTRINSIC_DEPTH: u32 = 256;
+
+/// Transaction info for txpool RPCs with Option fields to match Anvil's null values
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TxpoolTransactionInfo {
+    pub hash: H256,
+    pub block_hash: Option<H256>,
+    pub block_number: Option<sp_core::U256>,
+    pub transaction_index: Option<sp_core::U256>,
+    pub from: sp_core::H160,
+    pub transaction_signed: TransactionSigned,
+}
+
+/// Decode extrinsic into ETH transaction payload and signed transaction
+pub(super) fn decode_eth_transaction(
+    tx_data: &Arc<polkadot_sdk::sp_runtime::OpaqueExtrinsic>,
+) -> Option<(Vec<u8>, TransactionSigned)> {
+    let encoded = tx_data.encode();
+    let ext =
+        UncheckedExtrinsic::decode_all_with_depth_limit(MAX_EXTRINSIC_DEPTH, &mut &encoded[..])
+            .ok()?;
+
+    let polkadot_sdk::sp_runtime::generic::UncheckedExtrinsic {
+        function: RuntimeCall::Revive(polkadot_sdk::pallet_revive::Call::eth_transact { payload }),
+        ..
+    } = ext.0
+    else {
+        return None;
+    };
+
+    let signed_tx = TransactionSigned::decode(&payload).ok()?;
+
+    Some((payload, signed_tx))
+}
+
+/// Check if transaction matches ETH hash
+pub(super) fn transaction_matches_eth_hash(
+    tx_data: &Arc<polkadot_sdk::sp_runtime::OpaqueExtrinsic>,
+    target_eth_hash: B256,
+) -> bool {
+    let Some((payload, _signed_tx)) = decode_eth_transaction(tx_data) else {
+        return false;
+    };
+
+    let tx_eth_hash = keccak256(&payload);
+    B256::from_slice(tx_eth_hash.as_ref()) == target_eth_hash
+}
+
+/// Fields extracted from an Ethereum transaction
+pub(super) struct TransactionFields {
+    pub nonce: sp_core::U256,
+    pub to: Option<sp_core::H160>,
+    pub value: sp_core::U256,
+    pub gas: sp_core::U256,
+    pub gas_price: sp_core::U256,
+}
+
+/// Recover sender address from signed transaction as Substrate H160 type
+pub fn recover_sender_address(signed_tx: &TransactionSigned) -> Result<sp_core::H160, ()> {
```

Comment:
this is better suited in the host function module. and let's rename it to recover_maybe_impersonated_address to make its purpose more obvious

Reply:
or in the impersonation module maybe